### PR TITLE
Add payment circuit persistence

### PR DESCRIPTION
**Pull request by AndrewSamokhvalov**
In this commit the payment circuit persistence have been added which
is needed to properly propagate the settle/fail htlc messages after node
restart.

### CODE REVIEW

File: htlcswitch/circuit.go [hunk 1]
```diff
@@ -53,82 +80,220 @@ func newPaymentCircuit(src, dest lnwire.ShortChannelID, key circuitKey,
...
+
+// Decode reads the data from the byte stream and initialize the
+// payment circuit with data.
+func (c *paymentCircuit) Decode(r io.Reader) error {
+	if _, err := r.Read(c.PaymentHash[:]); err != nil {
+		return err
+	}
+
+	var srcBytes [8]byte
+	if _, err := r.Read(srcBytes[:]); err != nil {
```
Comment (PirosB3) [inline htlcswitch/circuit.go]: can we always assume `r` only contains 8 bytes? `r.Read()` will also return the number of types read (as first argument). Shouldn't we assert that it's 8?

File: htlcswitch/circuit.go [hunk 2]
```diff
@@ -53,82 +80,220 @@ func newPaymentCircuit(src, dest lnwire.ShortChannelID, key circuitKey,
...
+	c.Src = lnwire.NewShortChanIDFromInt(binary.BigEndian.Uint64(srcBytes[:]))
+
+	var destBytes [8]byte
+	if _, err := r.Read(destBytes[:]); err != nil {
+		return err
+	}
+	c.Dest = lnwire.NewShortChanIDFromInt(binary.BigEndian.Uint64(destBytes[:]))
+
+	var code [1]byte
+	if _, err := r.Read(code[:]); err != nil {
```
Comment (PirosB3) [inline htlcswitch/circuit.go]: You could use `ReadByte()`
```
func (b *Reader) ReadByte() (byte, error)
```

File: htlcswitch/circuit.go [hunk 3]
```diff
@@ -53,82 +80,220 @@ func newPaymentCircuit(src, dest lnwire.ShortChannelID, key circuitKey,
...
+		refCounter := uint64(1)
+		if data := counterBucket.Get(key[:]); data != nil {
+			refCounter = binary.BigEndian.Uint64(data)
+			refCounter++
+		}
+
+		// Update reference counter of the payment circuit.
+		var refCounterBytes [8]byte
+		binary.BigEndian.PutUint64(refCounterBytes[:], refCounter)
+		if err := counterBucket.Put(key[:], refCounterBytes[:]); err != nil {
```
Comment (PirosB3) [inline htlcswitch/circuit.go]: Just checking: this block of code is threadsafe right? From what I read `db.Batch` should be atomic, but not 100% convinced :)

File: htlcswitch/circuit.go [hunk 4]
```diff
@@ -53,82 +80,220 @@ func newPaymentCircuit(src, dest lnwire.ShortChannelID, key circuitKey,
...
-	return nil
+		return circuitBucket.Put(key[:], b.Bytes())
+	})
 }
 
-// remove destroys the target circuit by removing it from the circuit map.
-func (m *circuitMap) remove(key circuitKey) (*paymentCircuit, error) {
-	m.Lock()
-	defer m.Unlock()
+// remove destroys the target circuit by removing it from the circuit storage.
```
Comment (PirosB3) [inline htlcswitch/circuit.go]: looks like circuitKey doesn't get removed all the time right? ex. in the case `refCounter` isn't 0 on L263. Or should this never happen?

File: htlcswitch/circuit.go (hunk 1 above)
Comment (Roasbeef) [inline htlcswitch/circuit.go]: good point, readful should be used here and everywher else

**Comment by Roasbeef** [comment]
Replaced by #629.

**Pull request closed without merging**